test/evp_extra_test.c

```diff
@@ -1465,9 +1465,8 @@ static int test_EVP_PKEY_set1_DH(void)
             || !TEST_int_eq(EVP_PKEY_id(pkey1), EVP_PKEY_DHX))
         goto err;
 
-
     if(!TEST_true(EVP_PKEY_set1_DH(pkey2, pkcs3dh))
-            || !TEST_int_eq(EVP_PKEY_id(pkey2), EVP_PKEY_DH))
+            || !TEST_int_eq(EVP_PKEY_id(pkey2), EVP_PKEY_DHX))
```

Comment:
Errr,....say what??? I missed this before. Why is this changing? Surely pkcs3dh should have type EVP_PKEY_DH??

Reply:
DHX is the newer format that supports saving of q. If we load a named group then it really does have a known 'q' value. The code determines if it uses dhx by checking if it has a q value.
If you really want the old behaviour then a hacky dont use dhx flag would need to be added when we load by nid (or store the nid). Which way do you want?
Is this what we really want?

Reply:
The whole point of this test is check that if you assign a X9.42 DH object to an EVP_PKEY then you get an EVP_PKEY_DHX, whereas if you assign a PKCS.3 DH object (no q) then you get an EVP_PKEY_DH. After this PR it no longer tests that.

Reply:
I'm also a little concerned about the various explicit uses of EVP_PKEY_DH in libssl (search for `EVP_PKEY_DH`)

Reply:
I think until this PR none of the DH "named groups" that we used within a libssl context ever returned EVP_PKEY_DHX. On the up-side I think all of the DH "named group" code in libssl is only in master (not 1.1.1).

Reply:
> whereas if you assign a PKCS.3 DH object (no q) then you get an EVP_PKEY_DH. After this PR it no longer tests that.

Part of the issue is that TLS doesnt send 'q' so that is probably why it just uses the EVP_PKEY_DH type.
It is debatable whether this named group doesnt have a 'q' - because it does have a known q - we just werent setting it before.. So it looks like you want it to store some extra state and actually still use the old type for these groups (I will still store the q) but it wont save it as dhx..

I will think about this some more - maybe q is just a value required in some operations (such as printing and validation and we dont need to store it at all.. just calculate when required (or store into another variable or flag it)..

Reply:
IMO, if we have a q then we should use it.

I think there are two issues here:

1) The intention of the test was to check what happens when we assign DH objects without a q to an EVP_PKEY. It should end up with the EVP_PKEY_DH type - it no longer tests that

2) I have no problem with these named groups being EVP_PKEY_DHX type. We just need to review the libssl code to make sure it is still correct - because that code was written with the expectation that it would be EVP_PKEY_DH. I *think* all of the instances of this in libssl are related to TLSv1.3. The whole thing about q not being sent in TLS is irrelevant, because that only applies to TLSv1.2 and below. From TLSv1.3 we use named DH groups.

Reply:
> I *think* all of the instances of [named groups with EVP_PKEY_DH type] in libssl are related to TLSv1.3

I think if you support those curves you have to support them in 1.2 as well as 1.3, so it's not necessarily limited to TLS 1.3

Reply:
> I think if you support those curves you have to support them in 1.2 as well as 1.3, so it's not necessarily limited to TLS 1.3

Err...really? These groups are for DHE key exchange only. AFAIK TLSv1.2 lacks the ability to select a group by name in the ServerKeyExchange message for DHE - you can only provide the parameters explicitly.

Reply:
RFC 7919 specifies negotiated FFDHE named groups for TLS 1.0, 1.1, and 1.2

Reply:
Ah, but I see your point that the server still sends explicit parameters in ServerKeyExchange in that case, and the client's on the hook to validate them against the named group(s)

Reply:
We do not support RFC7919 (actually I wasn't aware of it until you pointed me at it). Looking at that RFC it still uses explicit parameters, i.e. it introduces named ffdhe groups in the "supported groups" extension,, but then the parameters themselves are still passed explicitly in the ServerKeyExchange. AFAIK there is nothing in the TLSv1.3 RFC that  requires us to start supporting RFC7919 in TLSv1.2. If we see the ffdhe groups in a TLSv1.2 connection we just ignore them.

Reply:
> AFAIK there is nothing in the TLSv1.3 RFC that requires us to start supporting RFC7919 in TLSv1.2.

Indeed, it is not listed in the "Updates Affecting TLS 1.2" section of RFC 8446

> If we see the ffdhe groups in a TLSv1.2 connection we just ignore them.

Do we want to make an issue to track changing that (to not ignore them)?

Reply:
#10971